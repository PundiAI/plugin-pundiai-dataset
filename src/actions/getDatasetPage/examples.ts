export default [
    
    [
        {
            user: "{{user1}}",
            content: {
                text: "Find image datasets in PNG format"
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Filtering image data...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "Show all datasets from 0x123..."
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Retrieving address datasets...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "Find video datasets larger than 10MB"
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Filtering large files...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "Show page 3 of public datasets, 100 items per page"
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Loading public data...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],

    [
        {
            user: "{{user1}}",
            content: {
                text: "Find multimodal datasets"
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Searching multimodal data...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "Datasets"
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Fetching default datasets...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "Show CSV files under 500KB"
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Filtering small CSV files...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "List datasets owned by 0xabc..."
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Retrieving owner datasets...",
                action: "GET_DATASET_PAGE",
            }
        }
    ],
    [
        {
            user: "{{user1}}",
            content: {
                text: "Find text datasets between 500KB and 2MB"
            }
        },
        {
            user: "{{agent}}",
            content: {
                text: "Filtering text datasets by size...",
                action: "GET_DATASET_PAGE",
            }
        }
    ]
]