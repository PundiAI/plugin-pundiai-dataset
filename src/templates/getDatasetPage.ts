export const getDatasetPageTemplate = `
## Purpose
Extract dataset search parameters from user queries with automatic validation and friendly guidance.
Given the most recent message only, extract information needed to perform dataset searches.


## Dataset Search Structure
A complete Search dataset object contains:
[
  {
    "contractAddress": "string (Ethereum address)",
     "tokenId": "string",
    "ownerAddress": "string (Ethereum address)",
    "datasetName": "string (repository/name format)",
    "permissions": "string (public/private)",
    "hash": "string (file hash)",
    "fileSize": "number (bytes)",
    "updateDt": "number (timestamp)",
  }
]"

## Search Parameters Structure
A complete dataset search object contains:
{
  "datasetName": "",           // Dataset name or partial match (empty for all)
  "address": "",               // Ethereum address (42 chars, empty if not specified)
  "addressType": 0,            // 0=user, 1=owner, 2=either
  "datasetDataType": "",       // Data type: "text","image","audio","video","tabular","multimodal"
  "datasetFormat": "",         // File format: ".csv",".json",".txt",".png",etc.
  "fileSizeStart": 0,          // Min size in bytes (default 0)
  "fileSizeEnd": 1000000,      // Max size in bytes (default 1MB)
  "permissions": "",           // "public","private" or empty for both
  "pageIndex": 1,              // Page number (1-based)
  "pageSize": 50               // Items per page (default 50)
}

## Extraction Rules
1. For basic queries, only extract explicitly mentioned fields with defaults
2. If user requests advanced search, extract all available filters
3. Return default values for unspecified fields
4. Validate address formats and enum values

## Examples
1. For "Find public datasets":
\`\`\`json
{
  "permissions": "public",
  "pageIndex": 1,
  "pageSize": 50
}
\`\`\`

2. For comprehensive search request:
\`\`\`json
{
  "datasetName": "finance",
  "address": "0xEF5c432D6FE9Da32ab71990F0133fa8a858B1DDf",
  "addressType": 1,
  "datasetDataType": "tabular",
  "datasetFormat": ".csv",
  "fileSizeStart": 10000,
  "fileSizeEnd": 500000,
  "permissions": "public",
  "pageIndex": 1,
  "pageSize": 20
}
\`\`\`

{{recentMessages}}
Analyze the LAST message only and extract search parameters based on:
1. The specificity of the request (basic or advanced search)
2. The available filters in the message
3. Return only the fields that can be determined with defaults
4. Provide validation errors for invalid inputs when appropriate`;