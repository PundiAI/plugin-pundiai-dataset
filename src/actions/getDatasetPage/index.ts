import { Action, HandlerCallback, IAgentRuntime, Memory, State, elizaLogger, composeContext, ModelClass, generateObjectDeprecated } from "@elizaos/core";
import examples from "./examples.ts";
import { getV1ViewQueryDatasetPage } from "../../utils/service"
import { GetV1ViewQueryDatasetPageQuery } from "../../utils/service/index.type.ts"
import { getDatasetPageTemplate } from "../../templates/getDatasetPage.ts"

export const getDatasetPage: Action = {
    name: "GET_DATASET_PAGE",
    description: "Get information about the dataset",
    similes: [],
    examples: examples,
    validate: async (_runtime: IAgentRuntime, message: Memory) => {
        return true
    },
    handler: async (runtime: IAgentRuntime, message: Memory, state?: State, _options?: { [key: string]: unknown; }, callback?: HandlerCallback): Promise<boolean> => {
        elizaLogger.log("Starting Moralis GET_DATASET_PAGE handler...");

        let currentState: State;
        if (!state) {
            currentState = (await runtime.composeState(message)) as State;
        } else {
            currentState = await runtime.updateRecentMessageState(state);
        }

        try {
            const statsContext = composeContext({
                state: currentState,
                template: getDatasetPageTemplate,
            });
            const content = (await generateObjectDeprecated({
                runtime,
                context: statsContext,
                modelClass: ModelClass.LARGE,
            })) as unknown as GetV1ViewQueryDatasetPageQuery;

            if (!content || typeof content !== "object") {
                throw new Error("Invalid response format from model");
            }

            const params = {
                ...(content.datasetName && { datasetName: content.datasetName }),
                ...(content.address && { address: content.address }),
                ...(content.addressType && { 
                  addressType: Number(content.addressType) 
                }),
                ...(content.datasetDataType && { 
                  datasetDataType: content.datasetDataType 
                }),
                ...(content.datasetFormat && { 
                  datasetFormat: content.datasetFormat 
                }),
                ...(content.permissions && { 
                  permissions: content.permissions 
                }),
                ...(content.fileSizeStart && { 
                    fileSizeStart: Number(content.fileSizeStart)
                  }),
                  ...(content.fileSizeEnd && { 
                    fileSizeEnd: Number(content.fileSizeEnd)
                  }),
                pageIndex: 1, 
                pageSize: 20,
              };

            elizaLogger.log("Request params:", params);

            const response = await getV1ViewQueryDatasetPage(params)

            const datasetData = response.data;

            if (callback) {

                if (datasetData.data.length <= 0){
                    callback({
                        text: "Here is the dataset not data.",
                        content: {
                            params,
                            datasetData
                        },
                    });
                    return true;
                }

                callback({
                    text: "Here is the dataset data: \n" + JSON.stringify(datasetData, null, 2),
                    content: {
                        params,
                        datasetData
                    },
                });
            }

            return true;
        } catch (error: unknown) {
            elizaLogger.error("Error in GET_DATASET_PAGE handler:", error);
            const errorMessage =
                error instanceof Error ? error.message : String(error);
            if (callback) {
                callback({
                    text: `Error fetching dataset data: ${errorMessage}`,
                    content: { error: errorMessage },
                });
            }

            return false;
        }
    }
}

