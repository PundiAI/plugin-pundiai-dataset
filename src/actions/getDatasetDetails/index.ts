import { Action, HandlerCallback, IAgentRuntime, Memory, State, elizaLogger, composeContext, ModelClass, generateObjectDeprecated } from "@elizaos/core";
import examples from "./examples.ts";
import { getV1ViewQueryDatasetDetails } from "../../utils/service"
import { GetV1ViewQueryDatasetDetailsQuery } from "../../utils/service/index.type.ts"
import { getDatasetDetailsTemplate } from "../../templates/getDatasetDetails";

export const getDatasetDetails: Action = {
    name: "GET_DATASET_DETAILS",
    description: "Get data set details by token id and contract address",
    similes: [],
    examples: examples,
    validate: async (_runtime: IAgentRuntime, message: Memory) => {
        return true
    },
    handler: async (runtime: IAgentRuntime, message: Memory, state?: State, _options?: { [key: string]: unknown; }, callback?: HandlerCallback): Promise<boolean> => {
        elizaLogger.log("Starting Moralis GET_DATASET_DETAILS handler...");

        let currentState: State;
        if (!state) {
            currentState = (await runtime.composeState(message)) as State;
        } else {
            currentState = await runtime.updateRecentMessageState(state);
        }

        try {
            const statsContext = composeContext({
                state: currentState,
                template: getDatasetDetailsTemplate,
            });
            const content = (await generateObjectDeprecated({
                runtime,
                context: statsContext,
                modelClass: ModelClass.LARGE,
            })) as unknown as GetV1ViewQueryDatasetDetailsQuery;

            if (!content || typeof content !== "object") {
                throw new Error("Invalid response format from model");
            }

            if (!content.tokenId || content.tokenId === 'null') {
                throw new Error("Please provide the Token ID for this contract");
            }

            if (!content.contractAddress || content.contractAddress === 'null') {
                throw new Error("Please provide the contract address");
            }

            const params = {
                tokenId: content.tokenId ,
                contractAddress: content.contractAddress
            };

            elizaLogger.log("Request params:", params);

            const response = await getV1ViewQueryDatasetDetails(params)

            const detailsData = response.data;

            if (callback) {
                callback({
                    text: "Here are the dataset details: \n" + JSON.stringify(detailsData, null, 2),
                    content: {
                        params,
                        detailsData
                    },
                });
            }

            return true;
        } catch (error: unknown) {
            elizaLogger.error("Error in GET_DATASET_DETAILS handler:", error);
            const errorMessage =
                error instanceof Error ? error.message : String(error);
            if (callback) {
                callback({
                    text: `Error fetching dataset details: ${errorMessage}`,
                    content: { error: errorMessage },
                });
            }

            return false;
        }
    }
}

