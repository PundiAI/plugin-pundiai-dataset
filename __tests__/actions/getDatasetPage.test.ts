import { describe, it, expect, vi, beforeEach, } from 'vitest';
import { elizaLogger, ModelClass, generateObjectDeprecated, composeContext } from '@elizaos/core';
import { getDatasetPage } from "../../src/actions/getDatasetPage";

describe('getDatasetPage action', () => {
    const mockRuntime = {
        composeState: vi.fn(),
        updateRecentMessageState: vi.fn(),
        getPluginConfig: vi.fn(),
    };

    const mockMessage = {};
    const mockState = {};
    const mockCallback = vi.fn();

    beforeEach(() => {
        vi.clearAllMocks();

        // Mock runtime functions
        mockRuntime.composeState.mockResolvedValue(mockState);
        mockRuntime.updateRecentMessageState.mockResolvedValue(mockState);
        mockRuntime.getPluginConfig.mockResolvedValue({});

        // Mock the core functions
        vi.mocked(elizaLogger.log).mockImplementation(() => {});
        vi.mocked(elizaLogger.error).mockImplementation(() => {});
        vi.mocked(elizaLogger.success).mockImplementation(() => {});
        vi.mocked(composeContext).mockReturnValue({});
    });

    it('should fetch and format dataset data', async () => {

        vi.mocked(generateObjectDeprecated).mockResolvedValueOnce({
            "datasetName": "May.6.George",
            "pageIndex": 1,
            "pageSize": 20
          });

        await getDatasetPage.handler(mockRuntime, mockMessage, mockState, {}, mockCallback);

        expect(mockCallback).toMatchSnapshot();
    })
});