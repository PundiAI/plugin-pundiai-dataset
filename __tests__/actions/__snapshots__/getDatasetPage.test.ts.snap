// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDatasetPage action > should fetch and format dataset data 1`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "content": {
          "datasetData": {
            "currentPageNo": 1,
            "data": [
              {
                "contractAddress": "0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
                "datasetName": "May.6.George",
                "fileSize": 0,
                "hash": "",
                "ownerAddress": "0x5d43dEDC405183DAD609c27BF7e7B0f2225A2B22",
                "permissions": "public",
                "tokenId": "17764",
                "updateDt": 1746498788217,
              },
            ],
            "pageSize": 20,
            "totalCount": 1,
            "totalPage": 1,
          },
          "params": {
            "datasetName": "May.6.George",
            "pageIndex": 1,
            "pageSize": 20,
          },
        },
        "text": "Here is the dataset data: 
{
  "data": [
    {
      "contractAddress": "0xbE0f332E99a557DB16E0a4a827f873f0A03d6CfE",
      "tokenId": "17764",
      "ownerAddress": "0x5d43dEDC405183DAD609c27BF7e7B0f2225A2B22",
      "datasetName": "May.6.George",
      "permissions": "public",
      "hash": "",
      "fileSize": 0,
      "updateDt": 1746498788217
    }
  ],
  "currentPageNo": 1,
  "totalCount": 1,
  "totalPage": 1,
  "pageSize": 20
}",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
